### PR TITLE
init auto loss scaling

### PR DESCRIPTION
**Pull request by Jie-Fang**
add auto loss scaling.

Usage:

1. Cast specific inputs to float16
```
imgs = fluid.layers.cast(images, "float16")
```
2. Decorate optimizer and minimize scaled loss
```
optimizer = fluid.optimizer.Adam(learning_rate=0.001)
	
 mp_optimizer = fluid.contrib.mixed_precision.decorate(
	            optimizer=optimizer, init_loss_scaling=8.0, use_dynamic_loss_scaling=True)
	
 scaled_loss, _, _ = mp_optimizer.minimize(avg_cost)
```

In the decorate function, you can set `incr_every__n_steps`, `decr_every_n_nan_or_inf`, `incr_ratio`, and `decr_ratio` to control the auto loss scaling details.

### CODE REVIEW

**Comment by kuke** [comment]
https://github.com/PaddlePaddle/Paddle/blob/7a3bb061d8289f38be074da05395bc729eb58035/paddle/fluid/API.spec#L32

**Comment by kuke** [comment]
https://github.com/PaddlePaddle/Paddle/blob/7a3bb061d8289f38be074da05395bc729eb58035/paddle/fluid/API.spec#L32 

Seems that you need to regenerate the API spec

**Comment by Jie-Fang** [comment]
> https://github.com/PaddlePaddle/Paddle/blob/7a3bb061d8289f38be074da05395bc729eb58035/paddle/fluid/API.spec#L32
> 
> Seems that you need to regenerate the API spec

I don't change `DistributeTranspilerConfig` API. When I  run `python tools/print_signatures.py paddle.fluid,paddle.reader > paddle/fluid/API.spec` command, it adds other info behind `paddle.fluid.DistributeTranspilerConfig.__init__`. Should I delete these info manually?

**Comment by kuke** [comment]
I regenerated the API spec, and it is right now.

**Review by shanyi15 — APPROVED** [review]

**Review by chengduoZH — APPROVED** [review]
